src/coordination/coordinator_log_store.cpp

```diff
@@ -13,229 +13,164 @@
-    disk_emul_logs_being_written_[cur_time + disk_emul_delay * 1000] = index;
-
-    // Remove entries greater than `index`.
-    auto entry = disk_emul_logs_being_written_.begin();
-    while (entry != disk_emul_logs_being_written_.end()) {
-      if (entry->second > index) {
-        entry = disk_emul_logs_being_written_.erase(entry);
-      } else {
-        entry++;
-      }
-    }
-    // disk_emul_ea_.invoke();
+  uint64_t next_slot{0};
+  {
+    auto lock = std::lock_guard{logs_lock_};
+    next_slot = start_idx_ + logs_.size() - 1;
+    logs_[next_slot] = clone;
   }
+  return next_slot;
 }
 
-ptr<std::vector<ptr<log_entry>>> CoordinatorLogStore::log_entries(ulong start, ulong end) {
-  ptr<std::vector<ptr<log_entry>>> ret = cs_new<std::vector<ptr<log_entry>>>();
+void CoordinatorLogStore::write_at(uint64_t index, ptr<log_entry> &entry) {
+  ptr<log_entry> clone = MakeClone(entry);
 
-  ret->resize(end - start);
-  ulong cc = 0;
-  for (ulong ii = start; ii < end; ++ii) {
-    ptr<log_entry> src = nullptr;
-    {
-      std::lock_guard<std::mutex> l(logs_lock_);
-      auto entry = logs_.find(ii);
-      if (entry == logs_.end()) {
-        entry = logs_.find(0);
-        assert(0);
-      }
-      src = entry->second;
+  // Discard all logs equal to or greater than `index.
+  {
+    auto lock = std::lock_guard{logs_lock_};
+    auto itr = logs_.lower_bound(index);
+    while (itr != logs_.end()) {
+      itr = logs_.erase(itr);
     }
-    (*ret)[cc++] = MakeClone(src);
+    logs_[index] = clone;
   }
-  return ret;
 }
 
-// NOLINTNEXTLINE(google-default-arguments)
-ptr<std::vector<ptr<log_entry>>> CoordinatorLogStore::log_entries_ext(ulong start, ulong end,
-                                                                      int64 batch_size_hint_in_bytes) {
-  ptr<std::vector<ptr<log_entry>>> ret = cs_new<std::vector<ptr<log_entry>>>();
-
-  if (batch_size_hint_in_bytes < 0) {
-    return ret;
-  }
+ptr<std::vector<ptr<log_entry>>> CoordinatorLogStore::log_entries(uint64_t start, uint64_t end) {
+  auto ret = cs_new<std::vector<ptr<log_entry>>>();
+  ret->resize(end - start);
 
-  size_t accum_size = 0;
-  for (ulong ii = start; ii < end; ++ii) {
+  for (uint64_t i = start, curr_index = 0; i < end; ++i, ++curr_index) {
     ptr<log_entry> src = nullptr;
     {
-      std::lock_guard<std::mutex> l(logs_lock_);
-      auto entry = logs_.find(ii);
-      if (entry == logs_.end()) {
-        entry = logs_.find(0);
-        assert(0);
+      auto lock = std::lock_guard{logs_lock_};
+      if (auto const entry = logs_.find(i); entry != logs_.end()) {
+        src = entry->second;
+      } else {
+        throw RaftCouldNotFindEntryException("Could not find entry at index {}", i);
       }
-      src = entry->second;
     }
-    ret->push_back(MakeClone(src));
-    accum_size += src->get_buf().size();
-    if (batch_size_hint_in_bytes && accum_size >= (ulong)batch_size_hint_in_bytes) break;
+    (*ret)[curr_index] = MakeClone(src);
```

Comment:
Is it safe to append element without holding the logs_lock_?

Reply:
we are appending to local var so yes